### PR TITLE
[mlir][vector] Move transpose with unit-dim to shape_cast pattern

### PR DESCRIPTION
**Pull request by c-rhodes**
Moved from lowering to canonicalization.

### CODE REVIEW

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 1]
```diff
@@ -5564,12 +5564,51 @@ class FoldTransposeCreateMask final : public OpRewritePattern<TransposeOp> {
...
+
+    SmallVector<int64_t> permutation;
+    transpOp.getTransp(permutation);
+
+    if (resType.getRank() == 2 &&
+        ((resType.getShape().front() == 1 &&
+          !resType.getScalableDims().front()) ||
+         (resType.getShape().back() == 1 &&
+          !resType.getScalableDims().back())) &&
+        permutation == ArrayRef<int64_t>({1, 0})) {
```
Comment (dcaballe) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:5595]: As a follow-up patch, I wonder if we could generalize this to n-D dimensions where 0 or 1 of them is != 1? If I'm not missing something, the permutation pattern itself shouldn't even matter for those cases?

**Review by dcaballe — APPROVED** [review]
Thanks a lot! This fixes one of the issues I'm hitting right now :)

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp (hunk 1 above)
Comment (c-rhodes) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp:5595]: Sounds good I'll look into it soon 👍

**Comment by MaheshRavishankar** [comment]
Based on https://github.com/llvm/llvm-project/pull/72105#issuecomment-1819800797 I would hope that we can revert the move to canonicalization at least since that does not give any user control. The revert is in https://github.com/MaheshRavishankar/llvm-project/commit/551134b8e18b9b9f132567fe27d8335b2ea0e478

**Comment by MaheshRavishankar** [comment]
> Thanks a lot! This fixes one of the issues I'm hitting right now :)

Thats unfortunate. Maybe we should make this be a separate pattern that is added where needed. It seems like it cannot be handled everywhere.

**Comment by antiagainst** [comment]
+1. I'm also missing the context of why going down vector shape cast is preferrable. Can somebody explain or give me some pointers?

**Comment by banach-space** [comment]
I'm sorry that this is causing issues

> I'm also missing the context of why going down vector shape cast is preferrable. Can somebody explain or give me some pointers?

This is important in the context of scalable vectors:
* LLVM can handle arrays of scalable vectors (which means that `vector<1x[4]xf32>` is supported), however
* LLVM cannot deal with "scalable" arrays of vectors (which means that `vector<[4]x1xf32>` is not supported).

This canonicalisation merely flips the dimension to make things super easy further down the compilation stack (`vector<1x[4]xf32>` --> `vector<[4]x1xf32>`).

> Maybe we should make this be a separate pattern that is added where needed.

Would you be able to make a specific suggestion? What's available to make sure that a pattern doesn't trigger for a particular target?

Rather than reverting this, I'd much prefer for us to "move" or refactor this, so that it's no longer problematic. Currently this unblocking us and solving an issue for @dcaballe , so there are benefits of keeping this in tree.

**Comment by antiagainst** [comment]
I think creating a `populate*Patterns()` to expose this pattern should work? There are lots of existing vectore related patterns in the codebase doing that. That way it gives more control for different CodeGen flows and you can explicitly call that in your flow. Things put in canonicalization is triggering everywhere as we don't have a way to control it.

**Pull request merged**